File: templates/python/runner.py
#!/usr/bin/env python3
"""Gota runner (Python) - YOUR code. This is the file you edit.

It declares the operation(s) to measure and plugs them into the harness. There is no
timing, batching, or JSON here: `register` receives a Bencher `b` and the buffer
`data`, and calls `b.bench(name, op)` once per operation. `gota.run` does the rest.

    python3 runner.py [buffer_bytes] [warmup_seconds] [measure_seconds]
"""

from __future__ import annotations

import gota

IMPL = "example"  # your implementation's name


def register(b: gota.Bencher, data: bytearray) -> None:
    # Replace with your real operation(s). Each op closes over `data` and runs the
    # work to measure. Call b.bench once per operation you want in the table.
    def example():
        for i in range(len(data)):
            data[i] ^= 0x5A

    b.bench("example", example)


if __name__ == "__main__":
    gota.run(IMPL, register)

File: templates/haskell/runner.hs
{-# LANGUAGE BangPatterns #-}

-- | Gota runner (Haskell) - YOUR code. This is the file you edit.
--
-- It declares the operation(s) to measure and plugs them into the harness. There is no
-- timing, batching, or JSON here: @run@ hands you a 'Bencher' and the buffer, and you
-- call @bench b name op@ once per operation.
--
-- The op is @Word64 -> Word64@: it takes the running accumulator and returns a result.
-- The harness forces and re-feeds that result, so the work can be neither deferred (left
-- a thunk) nor shared (computed once) — in Haskell, laziness is the sink hazard, and
-- seeding the op with the accumulator is how you defeat it. Read-only ops (hashes,
-- checksums, reductions) fit a 'BS.ByteString' buffer naturally; an in-place transform
-- would use a mutable buffer instead.
--
--     ghc -O2 runner.hs -o runner
--     ./runner [buffer_bytes] [warmup_seconds] [measure_seconds]
import Data.Bits (xor)
import Data.Word (Word64)
import Gota (bench, run)
import qualified Data.ByteString as BS

main :: IO ()
main = run "example" $ \b buf ->
  -- Replace with your real operation(s). This example folds the buffer with FNV-1a,
  -- seeded by the accumulator so each call does the full work over `buf`.
  bench b "example" $ \seed ->
    BS.foldl' (\acc x -> (acc `xor` fromIntegral x) * 0x100000001b3) seed buf
